packages/react/src/global.js

```diff
@@ -101,35 +98,24 @@ export let Global: React.AbstractComponent<
       let node: HTMLStyleElement | null = document.querySelector(
         `style[data-emotion="${key} ${serialized.name}"]`
       )
-
       if (cache.sheet.tags.length) {
         sheet.before = cache.sheet.tags[0]
       }
       if (node !== null) {
+        // clear the hash so this node won't be recognizable as rehydratable by other <Global/>s
+        node.setAttribute('data-emotion', key)
         sheet.hydrate([node])
```

Comment:
it turns out that `.hydrate` **moves** those global styles in `<head/>` anyway (but it doesn't remove them, even temporarily) so maybe those changes in the `<Global/>` were not needed at all as there is a high chance that moving styles behaves very similar to flushing+reinjecting 

there is also a possibility that it doesn't (I have not observed any problems with either approach), I feel though that the implementation is somewhat simpler now as it's easier to reason about a single effect than about two, so I would be probably in favor of keeping this implementation, thoughts? cc @mitchellhamilton

Reply:
I don't mind how it is implemented as long as it preserves the insert position it got when mounted which I don't believe we have a test for currently and I don't think the changed implementation does it correctly. https://codesandbox.io/s/still-leaf-invs2?file=/src/App.js

Reply:
Yes, the changed implementation doesn't cover this scenario correctly. However, if we add updates to the mix then the current stable release doesn't handle it correctly either: https://codesandbox.io/s/optimistic-browser-0inm1?file=/src/App.js

Scenario:
1. edit second input
2. toggle the first one twice

The premise of React is that the UI is always derived from state and this breaks that rule anyway. I'm not really sure if that's worth supporting unless we would figure out stable insertion points but I don't think we can (without complicating things too much).

Reply:
Yeah, I know that's not supported and I wouldn't have expected it to be but "The premise of React is that the UI is always derived from state and this breaks that rule anyway" isn't really accurate, preserving state between rerenders but not remounts is expected and just how things work in React e.g. https://codesandbox.io/s/jovial-greider-f4cgq. I don't think we should make the case you proposed work but we shouldn't regress from the current behavior.

Reply:
> preserving state between rerenders but not remounts is expected and just how things work in React

This one is different though as what gets reset is self-contained whereas `<Global/>` isn't and affects the outside world by rendering styles in a mount/remount/etc-order dependent fashion. I still stand by the "premise of React" being broken here anyway since for any given set of props, state, and context **values** the visual output should be the same.

That being said - I also don't see a point in supporting what I've showcased in the sandbox but I see how this could be used as an argument against supporting your use case. If things are not the same 100% of the time I don't think it's really worth supporting the 90% here as it gives a false sense of safety. But, at this point in time, it's not really worth bikeshedding this anyway as this would be sort of a breaking change so it's much better to just revert to the previous solution. I'm just wondering if it's worth implementing SSRed style preservation on top of that previous solution.

Reply:
I've reverted the merging of those 2 layout effects - although I have not added a test for the mentioned case because I'm not super convinced that this is something that we should support in the long run and I didn't quite have the time to add it right now. The use case seems very esoteric to me and we cannot easily produce stable order of global styles throughout the tree anyway.

For now, I've also ensured that the rehydrated global styles are not flushed and reinjected but I'm not super sure If this is required. I don't think we had any reports about this causing problems, so maybe we should just revert this small change?

Reply:
I don't mind about the flushing and reinjecting change.

(also, i agree about the preserving order in Global not being a thing we should care about, it's really just about not breaking things for people now)